### PR TITLE
Created the tagstore class and made edited persistent datastore to in…

### PR DESCRIPTION
**Pull request by masterford**
…clude writethrough for tags
Also edited the tag class to include conversation IDs

### CODE REVIEW

File: src/main/java/codeu/model/data/Tag.java [hunk 1]
```diff
@@ -27,6 +30,11 @@ public UUID getId() {
     return id;
   }
 
+  /** Returns the ID of the Conversation this Message belongs to. */
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/Tag.java]: fix comment to Tag

File: src/main/java/codeu/model/data/Tag.java [hunk 2]
```diff
@@ -27,6 +30,11 @@ public UUID getId() {
...
+  public UUID getConversationId() {
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/Tag.java]: can't a tag belong to multiple conversations?

File: src/main/java/codeu/model/store/basic/DefaultDataStore.java [hunk 3]
```diff
@@ -95,19 +99,35 @@ public boolean isValid() {
...
+    return tags;
+  }
+
   private void addRandomUsers() {
 
     List<String> randomUsernames = getRandomUsernames();
     Collections.shuffle(randomUsernames);
 
-    for (int i = 0; i < DEFAULT_USER_COUNT; i++) {
+    for (int i = 0; i < DEFAULT_MESSAGE_COUNT; i++) {
```
Comment (joycezhao2) [inline src/main/java/codeu/model/store/basic/DefaultDataStore.java]: why is this changed from User to Message counts?

File: src/main/java/codeu/model/store/persistence/PersistentDataStore.java [hunk 4]
```diff
@@ -178,4 +212,14 @@ public void writeThrough(Conversation conversation) {
...
+  /** Write a Tag object to the Datastore service. */
+  public void writeThrough(Tag tag) {
+    Entity tagEntity = new Entity("chat-tags");
+    tagEntity.setProperty("uuid", tag.getId().toString());
+    tagEntity.setProperty("conv_uuid", tag.getConversationId().toString());
+    tagEntity.setProperty("tag", tag.getTag());
+    tagEntity.setProperty("creation_time", tag.getCreationTime().toString());
+    datastore.put(tagEntity);
+  }
 }
```
Comment (joycezhao2) [inline src/main/java/codeu/model/store/persistence/PersistentDataStore.java]: whitespace!

File: src/test/java/codeu/model/store/basic/TagStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    Mockito.verify(mockPersistentStorageAgent).writeThrough(inputTag);
+  }
+
+  private void assertEquals(Tag expectedTag, Tag actualTag) {
+    Assert.assertEquals(expectedTag.getId(), actualTag.getId());
+    Assert.assertEquals(expectedTag.getConversationId(), actualTag.getConversationId());
+    Assert.assertEquals(expectedTag.getTag(), actualTag.getTag());
+    Assert.assertEquals(expectedTag.getCreationTime(), actualTag.getCreationTime());
+  }
+}
```
Comment (joycezhao2) [inline src/test/java/codeu/model/store/basic/TagStoreTest.java]: whitespace!

File: src/main/java/codeu/model/store/basic/DefaultDataStore.java (hunk 3 above)
Comment (masterford) [reply src/main/java/codeu/model/store/basic/DefaultDataStore.java]: Good catch, I had meant to use message count for addRandomTags and not addRandomUsers

File: src/main/java/codeu/model/data/Tag.java [hunk 6]
```diff
@@ -27,6 +30,11 @@ public UUID getId() {
     return id;
   }
 
+  /** Returns the ID of the Conversation this Tag belongs to. */
+  public UUID getConversationId() {
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/Tag.java]: Can't a tag belong to more than one conversation? So if I tag multiple conversations with the word "work", then there should be only one tag object that stores references to each conversation. 
So perhaps we should instead have a list of conversation Ids

File: src/main/java/codeu/model/store/basic/TagStore.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  public List<Tag> getTagsInConversation(UUID conversationId) {
+
+    List<Tag> tagsInConversation = new ArrayList<>();
+
+    for (Tag tag : tags) {
+      if (tag.getConversationId().equals(conversationId)) {
+        tagsInConversation.add(tag);
+      }
+    }
+
```
Comment (joycezhao2) [inline src/main/java/codeu/model/store/basic/TagStore.java]: too much whitespace!

File: src/main/java/codeu/model/store/basic/TagStore.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    return tagsInConversation;
+  }
+
+  /** Sets the List of Tags stored by this TagStore. */
+  public void setTags(List<Tag> tags) {
```
Comment (joycezhao2) [inline src/main/java/codeu/model/store/basic/TagStore.java]: what would this method be used for?

File: src/test/java/codeu/model/data/TagTest.java [hunk 9]
```diff
@@ -10,12 +10,14 @@
   @Test
   public void testCreate() {
     UUID id = UUID.randomUUID();
+    UUID conv_id = UUID.randomUUID();
     String tag = "test_tag";
     Instant creation = Instant.now();
 
-    Tag test_tag = new Tag(id, tag, creation);
+    Tag test_tag = new Tag(id, conv_id, tag, creation);
```
Comment (joycezhao2) [inline src/test/java/codeu/model/data/TagTest.java]: If conversations are stored in a list, this test should be updated

File: src/main/java/codeu/model/data/Tag.java [hunk 10]
```diff
@@ -27,6 +30,11 @@ public UUID getId() {
...
+    return conversation;
+  }
+
   /** Returns the content of this Tag. */
   public String getTag() {
     return tag;
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/Tag.java]: can you change the title of this method? to reflect that it gets the content of the tag and not the object itself

File: src/main/java/codeu/model/data/Tag.java [hunk 11]
```diff
@@ -38,11 +43,24 @@ public void addConversation(Conversation conversation) {
...
+  public Conversation getConversation() {
+    return conversation;
+  }
+
   /** Returns the ID of this Tag. */
   public UUID getId() {
     return id;
   }
 
+<<<<<<< HEAD
```
Comment (joycezhao2) [inline src/main/java/codeu/model/data/Tag.java]: can you resolve this merge conflict?

File: src/main/java/codeu/model/store/basic/TagStore.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    return loaded;
+  }
+
+  /** Access the current set of tags known to the application. */
+  public List<Tag> getAllTags() {
+    return tags;
+  }
+
+  /** Add a new Tag to the current set of tags known to the application. */
+  public void addTag(Tag tag) {
```
Comment (joycezhao2) [inline src/main/java/codeu/model/store/basic/TagStore.java]: oooh it would be helpful to also have a getTag method like getUser in UserStore
